buildSrc/src/main/java/org/elasticsearch/gradle/DockerUtils.java

```diff
@@ -0,0 +1,228 @@
+         */
+        public Result getLastCommand() {
+            return lastCommand;
+        }
+    }
+
+    // package-private for testing
+    /**
+     * Given the string that <code>docker --version</code> prints, extract the docker version from it.
+     * @param dockerVersion the version string to parse
+     * @return the Docker version
+     */
+    static String extractVersion(String dockerVersion) {
+        final Pattern dockerVersionPattern = Pattern.compile("Docker version (\\d+\\.\\d+\\.\\d+(?:-[a-zA-Z0-9]+)?), build [0-9a-f]{7,40}");
+
+        final Matcher matcher = dockerVersionPattern.matcher(dockerVersion);
+        if (!matcher.matches()) {
+            throw new IllegalArgumentException("Can't extract version from: [" + dockerVersion + "]");
+        }
+
+        return matcher.group(1);
+    }
+
+    // package-private for testing
+    /**
+     * Checks whether the supplied Docker version is &gt;= 17.05
+     * @param dockerVersion the version to check
+     * @return true if the version is high enough
+     */
+    static boolean checkVersion(String dockerVersion) {
+        final String[] majorMinor = dockerVersion.split("\\.");
+
+        return (Integer.parseInt(majorMinor[0]) > 17
+            || (Integer.parseInt(majorMinor[0]) == 17 && Integer.parseInt(majorMinor[1]) >= 5));
+    }
+
+    /**
+     * Runs a command and captures the exit code, standard output and standard error.
+     * @param args the command and any arguments to execute
+     * @return a object that captures the result of running the command. If an exception occurring
+     * while running the command, or the process was killed after reaching the 10s timeout,
+     * then the exit code will be -1.
+     */
+    private static Result runCommand(String... args) {
+        if (args.length == 0) {
+            throw new IllegalArgumentException("Cannot execute with no command");
+        }
+
+        final ProcessBuilder command = new ProcessBuilder().command(args);
+        try {
+            final Process process = command.start();
+
+            if (process.waitFor(10, TimeUnit.SECONDS) == false) {
+                if (process.isAlive()) {
+                    process.destroyForcibly();
+                }
+
+                Result result = new Result(-1,
+                    IOUtils.toString(process.getInputStream()),
+                    IOUtils.toString(process.getErrorStream()));
+
+                throw new IllegalStateException(
+                    "Timed out running shell command: " + command + "\n" +
+                        "Result:\n" + result
+                );
+            }
+
+            String stdout = IOUtils.toString(process.getInputStream());
+            String stderr = IOUtils.toString(process.getErrorStream());
+
+            return new Result(process.exitValue(), stdout, stderr);
+        } catch (Exception e) {
+            return new Result(-1, "", e.getMessage());
```

Comment:
This doesn't seem right. If we are getting some unexpected exception we shouldn't just return a `Result` as though this was what was returned from `docker`. We should rethrow here.

Reply:
Exceptions now propagate.